browser/brave_cosmetic_resources_tab_helper.cc

```diff
@@ -0,0 +1,457 @@
+    const std::vector<std::string>& classes,
+    const std::vector<std::string>& ids,
+    const std::vector<std::string>& exceptions) {
+  base::Optional<base::Value> hide_selectors =
+      g_brave_browser_process->ad_block_service()->HiddenClassIdSelectors(
+          classes, ids, exceptions);
+
+  base::Optional<base::Value> regional_selectors =
+      g_brave_browser_process->ad_block_regional_service_manager()
+          ->HiddenClassIdSelectors(classes, ids, exceptions);
+
+  base::Optional<base::Value> custom_selectors =
+      g_brave_browser_process->ad_block_custom_filters_service()
+          ->HiddenClassIdSelectors(classes, ids, exceptions);
+
+  if (hide_selectors && hide_selectors->is_list()) {
+    if (regional_selectors && regional_selectors->is_list()) {
+      for (auto i = regional_selectors->GetList().begin();
+           i < regional_selectors->GetList().end(); i++) {
+        hide_selectors->Append(std::move(*i));
+      }
+    }
+  } else {
+    hide_selectors = std::move(regional_selectors);
+  }
+
+  auto result_list = std::make_unique<base::ListValue>();
+  if (hide_selectors && hide_selectors->is_list()) {
+    result_list->Append(std::move(*hide_selectors));
+  }
+  if (custom_selectors && custom_selectors->is_list()) {
+    result_list->Append(std::move(*custom_selectors));
+  }
+
+  return result_list;
+}
+
+bool IsVettedSearchEngine(const std::string& host) {
+  for (size_t i = 0; i < g_vetted_search_engines.size(); i++) {
+    size_t found_pos = host.find(g_vetted_search_engines[i]);
+    if (found_pos != std::string::npos) {
+      size_t last_dot_pos = host.find(".", found_pos + 1);
+      if (last_dot_pos == std::string::npos) {
+        return false;
+      }
+      if (host.find(".", last_dot_pos + 1) == std::string::npos) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+}  // namespace
+
+BraveCosmeticResourcesTabHelper::BraveCosmeticResourcesTabHelper(
+    content::WebContents* contents)
+    : WebContentsObserver(contents),
+      enabled_1st_party_cf_filtering_(false),
+      weak_factory_(this) {
+  if (g_observing_script->empty()) {
+    *g_observing_script = LoadDataResource(kCosmeticFiltersGenerated[0].value);
+  }
+}
+
+BraveCosmeticResourcesTabHelper::~BraveCosmeticResourcesTabHelper() = default;
+
+void BraveCosmeticResourcesTabHelper::GetUrlCosmeticResourcesOnUI(
+    content::GlobalFrameRoutingId frame_id,
+    const std::string& url,
+    bool do_non_scriptlets,
+    std::unique_ptr<base::ListValue> resources) {
+  if (!resources) {
+    return;
+  }
+  for (auto& elem : resources->GetList()) {
+    base::DictionaryValue* resources_dict;
+    if (!elem.GetAsDictionary(&resources_dict)) {
+      continue;
+    }
+    std::string pre_init_script;
+    std::string scriptlet_init_script;
+    std::string non_scriptlet_init_script;
+    std::string json_to_inject;
+    if (base::JSONWriter::Write(*(resources_dict->FindPath("injected_script")),
```

Comment:
btw I woudl suggest moving as much as we can from the UI. (I have little clue about how large those json/values could be, but we are doing it pretty frequently)

Reply:
I could be several thousand Kb, but it's not for all websites, most websites don't need that. Not sure also switching context here and there will be better, but I'm creating a follow up to check that

Reply:
several thousands Kb means several Mb? I think it blocks us, because the UI performance will significantly degrade. 
We already have code that works on a TaskRunner, so should be easy to move there as much code as possible

Reply:
@iefremov sorry I meant hundred, not sure why put thousand, anyways that particular place is in production already on Android since summer https://github.com/brave/brave-core/blob/80b445649561388070efadf65737872fbeb8a70c/browser/android/brave_cosmetic_resources_tab_helper.cc#L79. Here is an example of scripts https://github.com/brave/brave-core/commit/80b445649561388070efadf65737872fbeb8a70c#diff-6ee12903241cbc4b7cb884063efb2a62230f78f7f86c3bdddbfd09a3af84a83aL38

Reply:
then we are probably fine for now. Would you mind adding `SCOPED_UMA_HISTOGRAM_TIMER("Brave.CosmeticFilters. GetUrlCosmeticResourcesOnUI");` to this method and checking the histogram on `brave://histograms`? And adding similar timers for other methods